File: day11/python/hex.py
#!/usr/bin/python3

import sys

def countSteps(steps):
    steps_N = steps.count('n')
    steps_NE = steps.count('ne')
    steps_SE = steps.count('se')
    steps_S = steps.count('s')
    steps_SW = steps.count('sw')
    steps_NW = steps.count('nw')

    # REMOVE OPPOSITE STEPS
    while steps_N > 0 and steps_S > 0:
        steps_N -= 1
        steps_S -= 1

    while steps_NE > 0 and steps_SW > 0:
        steps_NE -= 1
        steps_SW -= 1

    while steps_NW > 0 and steps_SE > 0:
        steps_NW -= 1
        steps_SE -= 1

    # COMBINE STEPS
    while steps_SW > 0 and steps_SE > 0:
        steps_SW -= 1
        steps_SE -= 1
        steps_S += 1

    while steps_NW > 0 and steps_NE > 0:
        steps_NW -= 1
        steps_NE -= 1
        steps_N += 1

    while steps_NE > 0 and steps_S > 0:
        steps_NE -= 1
        steps_S -= 1
        steps_SE += 1

    while steps_SE > 0 and steps_N > 0:
        steps_SE -= 1
        steps_N -= 1
        steps_NE += 1

    while steps_NW > 0 and steps_S > 0:
        steps_NW -= 1
        steps_S -= 1
        steps_SW += 1

    while steps_SW > 0 and steps_N > 0:
        steps_SW -= 1
        steps_N -= 1
        steps_NW += 1

    return steps_N + steps_NE + steps_SE + steps_S + steps_SW + steps_NW

if len(sys.argv) != 2:
    print('Provide input file as command line argument!')
else:
    steps = (open(sys.argv[1]).read())[:-1].split(',')

    print(countSteps(steps))

    distances = []
    for i in range(0, len(steps)):
        distances.append(countSteps(steps[:i]))

    print(max(distances))

File: day11/rust/hex.rs
use std::cmp::min;
use std::fs;

fn count_steps(steps: &Vec<&str>) -> u32 {
    let (mut n, mut ne, mut se, mut s, mut sw, mut nw) = (0, 0, 0, 0, 0, 0);
    for step in steps {
        match *step {
            "n" => n += 1,
            "ne" => ne += 1,
            "s" => s += 1,
            "se" => se += 1,
            "nw" => nw += 1,
            "sw" => sw += 1,
            _ => panic!("unknown direction!"),
        }
    }

    let mut steps: u32;

    // remove opposite steps
    steps = min(n, s);
    n -= steps;
    s -= steps;

    steps = min(ne, sw);
    ne -= steps;
    sw -= steps;

    steps = min(nw, se);
    nw -= steps;
    se -= steps;

    // combine steps
    steps = min(sw, se);
    sw -= steps;
    se -= steps;
    s += steps;

    steps = min(nw, ne);
    nw -= steps;
    ne -= steps;
    n += steps;

    steps = min(ne, s);
    ne -= steps;
    s -= steps;
    se += steps;

    steps = min(se, n);
    se -= steps;
    n -= steps;
    ne += steps;

    steps = min(nw, s);
    nw -= steps;
    s -= steps;
    sw += steps;

    steps = min(sw, n);
    sw -= steps;
    n -= steps;
    nw += steps;

    n + ne + se + s + sw + nw
}

fn main() {
    let mut hex = fs::read_to_string("../input").expect("file not found");
    hex = hex.trim().to_string();
    let steps = hex.split(',').collect();

    println!("{}", count_steps(&steps));

    let mut distances = Vec::new();
    for i in 0..steps.len() {
        distances.push(count_steps(&steps[0..i].to_vec()));
    }

    println!("{}", distances.iter().max().unwrap());
}
